src/ucs/sys/topo.c

```diff
@@ -158,53 +159,100 @@ ucs_status_t ucs_topo_get_device_bus_id(ucs_sys_device_t sys_dev,
+    ucs_topo_bus_id_str(&ucs_topo_global_ctx.devices[sys_dev].bus_id, 0,
+                        link_path + prefix_length, PATH_MAX - prefix_length);
+    if (realpath(link_path, path) == NULL) {
+        status = UCS_ERR_IO_ERROR;
+        goto out_unlock;
+    }
+
+    status = UCS_OK;
+
+out_unlock:
+    ucs_spin_unlock(&ucs_topo_global_ctx.lock);
+out:
+    return status;
+}
+
+static int ucs_topo_is_sys_root(const char *path)
+{
+    return !strcmp(path, UCS_TOPO_SYSFS_DEVICES_ROOT);
+}
+
+static int ucs_topo_is_pci_root(const char *path)
+{
+    int count;
+    sscanf(path, UCS_TOPO_SYSFS_DEVICES_ROOT "/pci%*d:%*d%n", &count);
+    return count == strlen(path);
 }
 
 ucs_status_t ucs_topo_get_distance(ucs_sys_device_t device1,
                                    ucs_sys_device_t device2,
                                    ucs_sys_dev_distance_t *distance)
 {
-    char path1[PATH_MAX], path2[PATH_MAX];
-    ssize_t path_distance;
+    char path1[PATH_MAX], path2[PATH_MAX], common_path[PATH_MAX];
+    size_t path_distance;
+    ucs_status_t status;
 
     /* If one of the devices is unknown, we assume near topology */
     if ((device1 == UCS_SYS_DEVICE_ID_UNKNOWN) ||
         (device2 == UCS_SYS_DEVICE_ID_UNKNOWN) || (device1 == device2)) {
-        goto default_distance;
+        *distance = ucs_topo_default_distance;
+        return UCS_OK;
     }
 
-    if ((device1 >= ucs_topo_num_devices()) ||
-        (device2 >= ucs_topo_num_devices())) {
-        return UCS_ERR_INVALID_PARAM;
+    status = ucs_topo_get_sysfs_path(device1, path1, sizeof(path1));
+    if (status != UCS_OK) {
+        return status;
     }
 
-    ucs_topo_get_bus_path(&ucs_topo_global_ctx.devices[device1].bus_id, path1,
-                          sizeof(path1));
-    ucs_topo_get_bus_path(&ucs_topo_global_ctx.devices[device2].bus_id, path2,
-                          sizeof(path2));
-
-    path_distance = ucs_path_calc_distance(path1, path2);
-    if (path_distance <= 0) {
-        /* Assume default distance for devices that cannot be found in sysfs */
-        goto default_distance;
+    status = ucs_topo_get_sysfs_path(device2, path2, sizeof(path2));
+    if (status != UCS_OK) {
+        return status;
     }
 
-    /* Rough approximation of bandwidth/latency as function of PCI distance in
-     * sysfs.
-     * TODO implement more accurate estimation, based on system type, PCIe
-     * switch, etc.
-     */
-    distance->latency   = 100e-9 * path_distance;
-    distance->bandwidth = (20000 / path_distance) * UCS_MBYTE;
-    return UCS_OK;
+    ucs_path_get_common_parent(path1, path2, common_path, sizeof(common_path));
+    if (ucs_topo_is_sys_root(common_path)) {
+        distance->latency = 300e-9;
+        switch (ucs_arch_get_cpu_model()) {
+        case UCS_CPU_MODEL_AMD_ROME:
+            distance->bandwidth = 5100 * UCS_MBYTE;
+            break;
+        default:
+            distance->bandwidth = 220 * UCS_MBYTE;
+            break;
+        }
```

Comment:
would it be better to have a separate function for this?

Reply:
done